File: app/src/main/java/network/minter/bipwallet/internal/data/CachedRepository.java
/*
 * Copyright (C) 2018 by MinterTeam
 * @link https://github.com/MinterTeam
 *
 * The MIT License
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package network.minter.bipwallet.internal.data;

import android.support.annotation.CallSuper;
import android.support.annotation.IntDef;
import android.support.annotation.Nullable;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.util.Date;
import java.util.concurrent.TimeUnit;

import io.reactivex.Observable;
import io.reactivex.Scheduler;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Action;
import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;
import io.reactivex.subjects.BehaviorSubject;
import timber.log.Timber;

/**
 * Wallet. 2017
 *
 * @author Eduard Maximovich <edward.vstock@gmail.com>
 */
public class CachedRepository<ResultModel, Entity extends CachedEntity<ResultModel>> {

    public final static int DEFAULT_EXPIRE_TIME = 30; //30 seconds
    protected final static Scheduler THREAD_MAIN = AndroidSchedulers.mainThread();
    protected final static Scheduler THREAD_IO = Schedulers.io();
    /**
     * Observer will call onNext(Action<ResultModel>) all time, even data is not expired and is
     * ready
     */
    protected final static int NOTIFY_ALWAYS_IF_EXISTS = 0;
    /**
     * Observer will call onNext(Action<ResultModel>) only if data is expired or force updated
     */
    protected final static int NOTIFY_ONLY_ON_UPDATE = 1;
    private final Entity mEntity;
    ResultModel mData;
    boolean mDataIsReady = false;
    Date mExpiredAt = null;
    Date mLastUpdateTime = null;
    private BehaviorSubject<MetaResult<ResultModel>> mMetaNotifier = BehaviorSubject.create();
    private BehaviorSubject<ResultModel> mNotifier = BehaviorSubject.create();
    private CompositeDisposable subscriptions = new CompositeDisposable();

    public CachedRepository(Entity entity) {
        mEntity = entity;
    }

    protected ResultModel initialData() {
        return mEntity.initialData();
    }

    public Class<Entity> getEntityClass() {
        return (Class<Entity>) mEntity.getClass();
    }

    public Entity getEntity() {
        return mEntity;
    }

    /**
     * @return Raw observable data
     */
    public Observable<ResultModel> getUpdateObservable() {
        return mEntity.getUpdatableData()
                .doOnNext(d -> {
                    Timber.tag(getClass().getSimpleName()).d(
                            "UpdateInternal: getUpdateObservable().onNext");
                    setData(d);
                })
                .doOnError(this::notifyOnError)
                .subscribeOn(THREAD_IO)
                .observeOn(THREAD_MAIN);
    }

    /**
     * Called after data updated
     */
    @CallSuper
    public void onAfterUpdate(ResultModel result) {
        mEntity.onAfterUpdate(result);
    }

    /**
     * Clear data and set data expired
     */
    public void clear() {
        mData = null;
        mDataIsReady = false;
        expire();
    }

    /**
     * Observer for update event
     *
     * @return obserable with ResultModel
     * @see ResultModel
     * @see Observable
     */
    public Observable<ResultModel> onUpdate() {
        return mNotifier
                .observeOn(THREAD_MAIN)
                .subscribeOn(THREAD_IO);
    }

    /**
     * Check data for expiration
     *
     * @return true if data is expired
     */
    public boolean isExpired() {
        if (mExpiredAt == null) {
            return true;
        }

        final Date now = new Date();
        final Date entityUpdatedAt = mExpiredAt;

        final boolean expired = now.getTime() >= entityUpdatedAt.getTime();
        if (expired) {
            onExpired();
        }

        return expired;
    }

    /**
     * Make data not expired
     */
    public void invalidateTime() {
        mLastUpdateTime = mExpiredAt = new Date();
        int expireTime = getExpireTime();
        expireTime *= 1000;
        mExpiredAt.setTime(mExpiredAt.getTime() + expireTime);
    }

    /**
     * Returns seconds to object life (ttl)
     *
     * @return integer seconds
     */
    public int getExpireTime() {
        return DEFAULT_EXPIRE_TIME;
    }

    /**
     * Make data expired
     */
    public void expire() {
        mLastUpdateTime = mExpiredAt = null;
    }

    @Nullable
    public Date getLastUpdate() {
        return mLastUpdateTime;
    }

    public void unsubscribe() {
        subscriptions.dispose();
        subscriptions.clear();
        mNotifier.onComplete();
        mNotifier = BehaviorSubject.create();
        mMetaNotifier.onComplete();
        mMetaNotifier = BehaviorSubject.create();
    }

    /**
     * @return True if data has set and not expired and not null
     */
    public boolean isDataReady() {
        return mDataIsReady && !isExpired() && getData() != null;
    }

    /**
     * Returns updatable data
     *
     * @return null if not loaded
     */
    public ResultModel getData() {
        return mData;
    }

    /**
     * Setting data
     *
     * @param data Any
     */
    @CallSuper
    public void setData(ResultModel data) {
        setDataInternal(data);
        onAfterUpdate(data);
    }

    /**
     * @param first
     * @return
     */
    public Observable<ResultModel> updateAfter(CachedRepository<ResultModel, Entity> first) {
        return updateAfter(first.mEntity.getUpdatableData());
    }

    /**
     * Creates zip function and after update first value, returns current object value
     *
     * @param first Any observable
     * @param <T>   Any type
     * @return Observable
     */
    public <T> Observable<ResultModel> updateAfter(Observable<T> first) {
        return Observable
                .zip(first, mEntity.getUpdatableData(), (firstT, resultModel) -> resultModel)
                .observeOn(THREAD_MAIN)
                .subscribeOn(THREAD_IO);
    }

    /**
     * @return
     * @see #observe()
     * @see #observeWithMeta()
     */
    @Deprecated
    public Observable<ResultModel> getOrUpdate() {
        if (isDataReady()) {
            return Observable.just(getData())
                    .observeOn(THREAD_MAIN)
                    .subscribeOn(THREAD_MAIN);
        }

        update();
        return onUpdate();
    }

    public void update() {
        updateInternal(false, null, null, null);
    }

    public void update(boolean force) {
        updateInternal(force, null, null, null);
    }

    public void update(boolean force, Consumer<ResultModel> onNext) {
        updateInternal(force, onNext, null, null);
    }

    public void update(boolean force, Consumer<ResultModel> onNext, Consumer<Throwable> onError) {
        updateInternal(force, onNext, onError, null);
    }

    public void update(boolean force, Consumer<ResultModel> onNext, Consumer<Throwable> onError,
                       Action onComplete) {
        updateInternal(force, onNext, onError, onComplete);
    }

    public void update(Consumer<ResultModel> onNext) {
        updateInternal(false, onNext, null, null);
    }

    public void update(Consumer<ResultModel> onNext, Consumer<Throwable> onError) {
        updateInternal(false, onNext, onError, null);
    }

    public void update(Consumer<ResultModel> onNext, Consumer<Throwable> onError,
                       Action onComplete) {
        updateInternal(false, onNext, onError, onComplete);
    }

    public void updateDelayed(int time, TimeUnit unit) {
        updateDelayedInternal(time, unit, false, null, null, null);
    }

    public void updateDelayed(int time, TimeUnit unit, boolean force) {
        updateDelayedInternal(time, unit, force, null, null, null);
    }

    public void updateDelayed(int time, TimeUnit unit, boolean force,
                              Consumer<ResultModel> onNext) {
        updateDelayedInternal(time, unit, force, onNext, null, null);
    }

    public void updateDelayed(int time, TimeUnit unit, boolean force, Consumer<ResultModel> onNext,
                              Consumer<Throwable> onError) {
        updateDelayedInternal(time, unit, force, onNext, onError, null);
    }

    public void updateDelayed(int time, TimeUnit unit, boolean force, Consumer<ResultModel> onNext,
                              Consumer<Throwable> onError, Action onComplete) {
        updateDelayedInternal(time, unit, force, onNext, onError, onComplete);
    }

    public void updateDelayed(int time, TimeUnit unit, Consumer<ResultModel> onNext) {
        updateDelayedInternal(time, unit, false, onNext, null, null);
    }

    public void updateDelayed(int time, TimeUnit unit, Consumer<ResultModel> onNext,
                              Consumer<Throwable> onError) {
        updateDelayedInternal(time, unit, false, onNext, onError, null);
    }

    public void updateDelayed(int time, TimeUnit unit, Consumer<ResultModel> onNext,
                              Consumer<Throwable> onError, Action onComplete) {
        updateDelayedInternal(time, unit, false, onNext, onError, onComplete);
    }

    /**
     * Return subscriber that subscribes to all updates of this repository including some additional
     * information about data
     * Every time, consumer will call #update(), subscriber will notified about updating
     *
     * @return ReplaySubject
     * @see MetaResult
     */
    public BehaviorSubject<MetaResult<ResultModel>> observeWithMeta() {
        return mMetaNotifier;
    }

    /**
     * Return subscriber that subscribes to all updates of this repository
     * Every time, consumer will call #update(), subscriber will notified about updating
     *
     * @return BehaviorSubject
     */
    public BehaviorSubject<ResultModel> observe() {
        return mNotifier;
    }

    final void setIsReadyInternal(boolean isReady) {
        mDataIsReady = isReady;
    }

    final void setDataInternal(ResultModel data) {
        mData = data;
        mDataIsReady = true;
        notifyOnSuccess(true);
    }

    protected void onExpired() {
    }

    @NotifyStrategy
    protected int getNotifyStrategy() {
        return NOTIFY_ALWAYS_IF_EXISTS;
    }

    protected synchronized CachedRepository<ResultModel, Entity> updateDelayedInternal(
            int delay, TimeUnit unit,
            boolean force,
            @Nullable Consumer<ResultModel> onNext,
            @Nullable Consumer<Throwable> onError,
            @Nullable Action onComplete) {

        Disposable sub = Observable.timer(delay, unit)
                .subscribeOn(THREAD_IO)
                .subscribe(
                        est -> updateInternal(force, onNext, onError, onComplete));

        subscriptions.add(sub);

        return this;
    }

    protected synchronized CachedRepository<ResultModel, Entity> updateInternal(
            boolean force,
            @Nullable Consumer<ResultModel> onSuccess,
            @Nullable Consumer<Throwable> onError,
            @Nullable Action onComplete
    ) {
        Observable<ResultModel> observable;

        if (!force && !isExpired() && isDataReady()) {
            if (getNotifyStrategy() == NOTIFY_ONLY_ON_UPDATE) {
                return this;
            }

            observable = Observable.just(getData())
                    .doOnNext(d -> {
                        Timber.tag(getClass().getSimpleName()).d("UpdateInternal: just().onNext (thread: %s)",
                                Thread.currentThread().getName());
                    })
                    .doOnError(this::notifyOnError)
                    .subscribeOn(THREAD_MAIN)
                    .observeOn(THREAD_MAIN);
        } else {
            observable = getUpdateObservable();
            invalidateTime();
        }

        observable
                .subscribe(
                        res -> {
                            Timber.d("OnNext with data: %s", res.toString());
                            callOnNext(onSuccess).accept(res);
                        },
                        callOnError(onError),
                        callOnComplete(onComplete)
                );

        return this;
    }

    protected void notifyOnSuccess(boolean isNew) {
        mMetaNotifier.onNext(new MetaResult<>(getData(), isNew));
        mNotifier.onNext(getData());
    }

    protected void notifyOnError(Throwable t) {
        mMetaNotifier.onError(t);
        mMetaNotifier = BehaviorSubject.create();
        mNotifier.onError(t);
        mNotifier = BehaviorSubject.create();
    }

    private Consumer<ResultModel> callOnNext(Consumer<ResultModel> chain) {
        return r -> {
            if (chain != null) {
                chain.accept(getData());
            }
        };
    }

    private Consumer<Throwable> callOnError(Consumer<Throwable> chain) {
        return t -> {
            if (chain != null) {
                chain.accept(t);
            }
        };
    }

    private Action callOnComplete(Action chain) {
        return () -> {
            if (chain != null) {
                chain.run();
            }
        };
    }

    @Retention(RetentionPolicy.SOURCE)
    @IntDef({NOTIFY_ONLY_ON_UPDATE, NOTIFY_ALWAYS_IF_EXISTS})
    @interface NotifyStrategy {
    }

    public final static class MetaResult<ResultModel> {
        private ResultModel data;
        private boolean isNew = false;

        MetaResult(ResultModel data, boolean isNew) {
            this.data = data;
            this.isNew = isNew;
        }

        public ResultModel value() {
            return data;
        }

        public boolean isNewData() {
            return isNew;
        }

        @Override
        public String toString() {
            return data.toString();
        }

        public Observable<MetaResult<ResultModel>> toObservable() {
            return Observable.just(this);
        }
    }

}

File: app/src/main/java/network/minter/bipwallet/internal/data/CachedRepository.kt
/*
 * Copyright (C) by MinterTeam. 2021
 * @link <a href="https://github.com/MinterTeam">Org Github</a>
 * @link <a href="https://github.com/edwardstock">Maintainer Github</a>
 *
 * The MIT License
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package network.minter.bipwallet.internal.data

import androidx.annotation.CallSuper
import io.reactivex.Observable
import io.reactivex.ObservableSource
import io.reactivex.Scheduler
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.exceptions.UndeliverableException
import io.reactivex.functions.Action
import io.reactivex.functions.Consumer
import io.reactivex.functions.Function
import io.reactivex.schedulers.Schedulers
import io.reactivex.subjects.BehaviorSubject
import network.minter.bipwallet.BuildConfig
import network.minter.bipwallet.internal.storage.KVStorage
import timber.log.Timber
import java.util.*
import java.util.concurrent.TimeUnit

/**
 * minter-android-wallet. 2018
 * @author Eduard Maximovich (edward.vstock@gmail.com)
 */
open class CachedRepository<ResultModel, Entity : CachedEntity<ResultModel>>(
        val storage: KVStorage,
        val entity: Entity
) {
    companion object {
        const val DEFAULT_EXPIRE_TIME = 30L //30 seconds
        protected val THREAD_MAIN: Scheduler = AndroidSchedulers.mainThread()
        protected val THREAD_IO = Schedulers.io()

        /**
         * Observer will call onNext(Action<ResultModel>) only if data is expired or force updated
         */
        protected const val NOTIFY_ONLY_ON_UPDATE = 1
        protected const val NOTIFY_ALWAYS_IF_EXISTS = 0

        private const val KEY_EXPIRED_ITEM = BuildConfig.MINTER_CACHE_VERS + "_expire_time_item_"
    }

    //    protected var mData: ResultModel
//    protected var mDataIsReady: Boolean
    protected var _expiredAt: Date? = null
    protected var lastUpdate: Date? = null
    private var mMetaNotifier: BehaviorSubject<MetaResult<ResultModel>> = BehaviorSubject.create()
    private var mNotifier: BehaviorSubject<ResultModel> = BehaviorSubject.create()
    private val subscriptions = CompositeDisposable()
    private val updateLock = Any()

    private var enableInMemoryCache: Boolean = false
    private var entityCachedData: ResultModel? = null

    private val expiredEntityStorageKey: String
        get() {
            return KEY_EXPIRED_ITEM + entity.getDataKey()
        }

    /**
     * Observer will call onNext(Action<ResultModel>) all time, even data is not expired and is ready
     */
    protected val notifyStrategy = NOTIFY_ALWAYS_IF_EXISTS

    /**
     * Returns seconds to object life (ttl)
     * @return integer seconds
     */
    var expireTime: Long = DEFAULT_EXPIRE_TIME
        private set

    val entityClass: Class<Entity>
        get() = entity.javaClass

    init {
        _expiredAt = storage[expiredEntityStorageKey, null]
        expire()
    }

    /**
     * @return Raw observable data
     */
    val updateObservable: Observable<ResultModel>
        get() {
            var observable = entity.getUpdatableData()

            if (retryWhenHandler != null) {
                observable = observable.retryWhen {
                    Timber.d("Call retryWhen for %s repo", entity.javaClass.simpleName)
                    retryWhenHandler!!.apply(it)
                }
            } else {
                observable = observable.doOnError { t: Throwable -> notifyOnError(t) }
            }

            observable = observable
                    .subscribeOn(THREAD_IO)

            return observable
        }

    private var retryWhenHandler: Function<Observable<out Throwable>, ObservableSource<*>>? = null

    fun retryWhen(handler: Function<Observable<out Throwable>, ObservableSource<*>>): CachedRepository<ResultModel, Entity> {
        retryWhenHandler = handler
        return this
    }

    /**
     * Called after data updated
     */
    @CallSuper
    fun onAfterUpdate(result: ResultModel) {
        entityCachedData = result
        entity.onAfterUpdate(result)
    }

    /**
     * Clear data and set data expired
     */
    fun clear() {
        entity.onClear()
        expire()
        mNotifier.onComplete()
        mNotifier = BehaviorSubject.create()
    }

    /**
     * Observer for update event
     * @return obserable with ResultModel
     * @see ResultModel
     *
     * @see Observable
     */
    fun onUpdate(): Observable<ResultModel> {
        return mNotifier
                .observeOn(THREAD_MAIN)
                .subscribeOn(THREAD_IO)
    }

    /**
     * Check data for expiration
     * @return true if data is expired
     */
    val isExpired: Boolean
        get() {
            if (_expiredAt == null) {
                Timber.d("Entity %s is expired=true", expiredEntityStorageKey)
                return true
            }
            val now = Date()
            val entityUpdatedAt: Date = _expiredAt!!
            val expired = now.time >= entityUpdatedAt.time
            if (expired) {
                onExpired()
            }
            Timber.d("Entity %s is expired=%b", expiredEntityStorageKey, expired)
            return expired
        }

    /**
     * Make data not expired
     */
    fun invalidateTime() {
        _expiredAt = Date()
        lastUpdate = _expiredAt
        var expireTime = expireTime
        expireTime *= 1000
        _expiredAt!!.time = _expiredAt!!.time + expireTime

        storage.put(expiredEntityStorageKey, _expiredAt)
    }

    fun setTimeToLive(ttlSeconds: Long): CachedRepository<ResultModel, Entity> {
        expireTime = ttlSeconds
        return this
    }

    fun setTimeToLive(time: Long, unit: TimeUnit): CachedRepository<ResultModel, Entity> {
        expireTime = unit.toSeconds(time)
        return this
    }

    fun setEnableInMemoryCache(enable: Boolean): CachedRepository<ResultModel, Entity> {
        enableInMemoryCache = enable
        return this
    }

    /**
     * Make data expired
     */
    fun expire() {
        _expiredAt = null
        lastUpdate = _expiredAt
        storage.delete(expiredEntityStorageKey)
    }

    fun unsubscribe() {
        subscriptions.dispose()
        subscriptions.clear()
        mNotifier.onComplete()
        mNotifier = BehaviorSubject.create()
        mMetaNotifier.onComplete()
        mMetaNotifier = BehaviorSubject.create()
    }

    /**
     * @return True if data has set and not expired and not null
     */
    val isDataReady: Boolean
        get() = entity.isDataReady() && !isExpired

    /**
     * Returns updatable data
     * @return null if not loaded
     */
    val data: ResultModel
        get() = getEntityData()

    /**
     * Setting data
     * @param data Any
     */
    @CallSuper
    fun notifyData(data: ResultModel) {
        onAfterUpdate(data)
        if (entity.onAfterUpdateDeferred(data) != null) {
            entity.onAfterUpdateDeferred(data)!!
                    .observeOn(Schedulers.io())
                    .subscribeOn(Schedulers.io())
                    .subscribe {
                        notifyOnSuccess(true)
                    }
        } else {
            notifyOnSuccess(true)
        }


    }

    /**
     * @param first
     * @return
     */
    fun updateAfter(first: CachedRepository<ResultModel, Entity>): Observable<ResultModel> {
        return updateAfter(first.entity.getUpdatableData())
    }

    /**
     * Creates zip function and after update first value, returns current object value
     * @param first Any observable
     * @param <T> Any type
     * @return Observable
     */
    @Suppress("NULLABLE_TYPE_PARAMETER_AGAINST_NOT_NULL_TYPE_PARAMETER")
    fun <T> updateAfter(first: Observable<T>): Observable<ResultModel> {
        return Observable
                .zip(first, entity.getUpdatableData(), { _: T, resultModel: ResultModel -> resultModel })
                .observeOn(THREAD_MAIN)
                .subscribeOn(THREAD_IO)
    }

    /**
     * Returns local data, or if doesn't loaded, returns observable to remote source
     * @see .observe
     * @see .observeWithMeta
     */
    fun fetch(): Observable<ResultModel> {
        if (isDataReady && !isExpired) {
            return Observable.just(getEntityData())
                    .subscribeOn(THREAD_IO)
                    .observeOn(THREAD_MAIN)
        }

        invalidateTime()
        return updateObservable
                .doOnNext(callOnNext(null))
                .doOnError(callOnError(null))
                .doOnComplete(callOnComplete(null))
    }

    @JvmOverloads
    fun update(
            force: Boolean,
            onNext: Consumer<ResultModel>? = null,
            onError: Consumer<Throwable>? = null,
            onComplete: Action? = null) {
        updateInternal(force, onNext, onError, onComplete)
    }

    @JvmOverloads
    fun update(
            onNext: Consumer<ResultModel>? = null,
            onError: Consumer<Throwable>? = null,
            onComplete: Action? = null) {
        updateInternal(false, onNext, onError, onComplete)
    }

    @JvmOverloads
    fun updateDelayed(
            time: Int,
            unit: TimeUnit = TimeUnit.SECONDS,
            force: Boolean = false,
            onNext: Consumer<ResultModel>? = null,
            onError: Consumer<Throwable>? = null,
            onComplete: Action? = null) {
        updateDelayedInternal(time, unit, force, onNext, onError, onComplete)
    }

    /**
     * Return subscriber that subscribes to all updates of this repository including some additional
     * information about data
     * Every time, consumer will call #update(), subscriber will notified about updating
     * @return ReplaySubject
     * @see MetaResult
     */
    fun observeWithMeta(): BehaviorSubject<MetaResult<ResultModel>> {
        return mMetaNotifier
    }

    /**
     * Return subscriber that subscribes to all updates of this repository
     * Every time, consumer will call #update(), subscriber will notified about updating
     * @return BehaviorSubject
     */
    fun observe(): BehaviorSubject<ResultModel> {
        return mNotifier
    }


    protected fun onExpired() {}

    @Synchronized
    protected fun updateDelayedInternal(
            delay: Int,
            unit: TimeUnit = TimeUnit.SECONDS,
            force: Boolean = false,
            onNext: Consumer<ResultModel>? = null,
            onError: Consumer<Throwable>? = null,
            onComplete: Action? = null
    ): CachedRepository<ResultModel, Entity> {
        val sub = Observable.timer(delay.toLong(), unit)
                .subscribeOn(THREAD_IO)
                .subscribe { updateInternal(force, onNext, onError, onComplete) }
        subscriptions.add(sub)
        return this
    }

    private fun getEntityData(): ResultModel {
        if (!enableInMemoryCache) {
            return entity.getData()
        }
        if (entityCachedData == null) {
            entityCachedData = entity.getData()
        }
        return entityCachedData ?: entity.getData()
    }

    fun updateBlockingSingle(force: Boolean = false,
                             onSuccess: Consumer<ResultModel>? = null,
                             onError: Consumer<Throwable>? = null,
                             onComplete: Action? = null): Single<ResultModel> {
        synchronized(updateLock) {
            val observable: Observable<ResultModel>
            if (!force && !isExpired && isDataReady) {
                observable = Observable.just(getEntityData())
            } else {
                observable = updateObservable
                invalidateTime()
            }

            return try {
                val res = observable.blockingFirst()
                callOnNext(onSuccess).accept(res)
                callOnComplete(onComplete).run()
                Single.just(res)
            } catch (t: Throwable) {
                callOnError(onError).accept(t)
                Single.error(t)
            }
        }
    }

    protected fun updateInternal(
            force: Boolean = false,
            onSuccess: Consumer<ResultModel>? = null,
            onError: Consumer<Throwable>? = null,
            onComplete: Action? = null
    ) {
        synchronized(updateLock) {
            val observable: Observable<ResultModel>
            if (!force && !isExpired && isDataReady) {
                if (notifyStrategy == NOTIFY_ONLY_ON_UPDATE) {
                    return
                }
                observable = Observable.just(getEntityData())
            } else {
                observable = updateObservable
                invalidateTime()
            }
            observable
                    .subscribe(
                            callOnNext(onSuccess),
                            callOnError(onError),
                            callOnComplete(onComplete)
                    )
        }
    }

    @Suppress("NULLABLE_TYPE_PARAMETER_AGAINST_NOT_NULL_TYPE_PARAMETER")
    protected fun notifyOnSuccess(isNew: Boolean) {
        mMetaNotifier.onNext(MetaResult(getEntityData(), isNew))
        mNotifier.onNext(getEntityData())
    }

    protected fun notifyOnError(t: Throwable?) {
        if (t == null) {
            Timber.w("Throwable after error is null")
            return
        }
        mMetaNotifier.onError(t)
        mMetaNotifier = BehaviorSubject.create()
        try {
            mNotifier.onError(t)
        } catch (e: UndeliverableException) {
        }
        mNotifier = BehaviorSubject.create()
    }

    private fun callOnNext(chain: Consumer<ResultModel>?): Consumer<ResultModel> {
        return Consumer {
            notifyData(it)
            chain?.accept(it)
        }
    }

    private fun callOnError(chain: Consumer<Throwable>?): Consumer<Throwable> {
        return Consumer {
            notifyOnError(it)
            chain?.accept(it)
        }
    }

    private fun callOnComplete(chain: Action?): Action {
        return Action {
            chain?.run()
        }
    }

    class MetaResult<ResultModel> internal constructor(private val data: ResultModel, isNew: Boolean) {
        var isNewData = false
        fun value(): ResultModel {
            return data
        }

        override fun toString(): String {
            return data.toString()
        }

        fun toObservable(): Observable<MetaResult<ResultModel>> {
            return Observable.just(this)
        }

        init {
            isNewData = isNew
        }
    }
}